ocrd/ocrd/mets_server.py

```diff
@@ -0,0 +1,237 @@
+"""
+# METS server functionality
+"""
+import re
+from os import environ, _exit
+from io import BytesIO
+from typing import Any, Dict, Optional, Union, List, Tuple
+
+from fastapi import FastAPI, Request, File, Form, UploadFile
+from fastapi.responses import JSONResponse
+from requests import request, Session as requests_session
+from requests_unixsocket import Session as requests_unixsocket_session
+from pydantic import BaseModel, Field, constr, ValidationError
+
+import uvicorn
+
+from ocrd_models import OcrdMets
+from ocrd_utils import initLogging, getLogger, deprecated_alias
+
+#
+# XXX HACKS TODO
+#
+initLogging()
+
+#
+# Models
+#
+
+class OcrdFileModel(BaseModel):
+    file_grp : str = Field()
+    file_id : str = Field()
+    mimetype : str = Field()
+    page_id : Union[str, None] = Field()
+    url : str = Field()
+
+class OcrdAgentModel(BaseModel):
+    name : str = Field()
+    _type : str = Field()
+    role : str = Field()
+    otherrole : str = Field()
+    othertype : str = Field()
+    notes : List[Tuple[Dict[str, str], Optional[str]]] = Field()
+
+class OcrdFileListModel(BaseModel):
+    files : List[OcrdFileModel] = Field()
+
+class OcrdFileGroupListModel(BaseModel):
+    file_groups : List[str] = Field()
+
+#
+# Client
+#
+
+class ClientSideOcrdFile:
+    """
+    Provides the same interface as :py:class:`ocrd_models.ocrd_file.OcrdFile`
+    but without attachment to :py:class:`ocrd_models.ocrd_mets.OcrdMets` since
+    this represents the response of the :py:class:`ocrd.mets_server.OcrdWorkspaceServer`.
+    """
+
+    def __init__(self, el, mimetype=None, pageId=None, loctype='OTHER', local_filename=None, mets=None, url=None, ID=None, fileGrp=None):
+        """
+        Args:
+            el (): ignored
+        Keyword Args:
+            mets (): ignored
+            mimetype (string): ``@MIMETYPE`` of this ``mets:file``
+            pageId (string): ``@ID`` of the physical ``mets:structMap`` entry corresponding to this ``mets:file``
+            loctype (string): ``@LOCTYPE`` of this ``mets:file``
+            local_filename (): ignored
+            url (string): ``@xlink:href`` of this ``mets:file``
+            ID (string): ``@ID`` of this ``mets:file``
+        """
+        self.ID = ID
+        self.mimetype = mimetype
+        self.url = url
+        self.loctype = loctype
+        self.pageId = pageId
+        self.fileGrp = fileGrp
+
+class ClientSideOcrdMets():
+    """
+    Replacement for :py:class:`ocrd_models.ocrd_mets.OcrdMets` with overrides for
+    :py:meth:`ocrd_models.ocrd_mets.OcrdMets.find_files`,
+    :py:meth:`ocrd_models.ocrd_mets.OcrdMets.find_all_files`, and
+    :py:meth:`ocrd_models.ocrd_mets.OcrdMets.add_file` to query via HTTP a
+    :py:class:`ocrd.mets_server.OcrdMetsServer`.
+    """
+
+    def __init__(self, host, port, socket):
+        self.log = getLogger('ocrd.workspace_client')
```

Comment:
You mean like `ocrd.workspace_client.uds` vs `ocrd.workspace_client.tcp`?

Reply:
Yes. Although I am not sure if that would help anyhow.

Reply:
Done  69dad92d7. It can't hurt, people can still override `ocrd.mets_client` to change both socket/host-port behavior.

Also renamed it `ocrd.mets_client`, `workspace_client` was from an earlier iteration of the PR.